stdlib/public/SDK/Foundation/NSDictionary.swift

```diff
@@ -138,6 +138,14 @@ extension Dictionary : _ObjectiveCBridgeable {
   }
 }
 
+extension NSDictionary : _HasCustomAnyHashableRepresentation {
+  // Must be @nonobjc to avoid infinite recursion during bridging
+  @nonobjc
+  public func _toCustomAnyHashable() -> AnyHashable? {
+    return AnyHashable(self as! Dictionary<AnyHashable, AnyHashable>)
```

Comment:
This isn't correct; the values might not be hashable.

Reply:
Also, don't forget NSSet?

Reply:
> This isn't correct; the values might not be hashable.

Oh, do you mean when they include `_SwiftValue`s whose corresponding Swift types aren't `Hashable`? Hm, that's a good point -- and I think it applies to `NSArray`, too. (And maybe even `NSDictionary` keys and `NSSet`.) 

I vaguely remember that this still ends up working, because we keep such values in their `AnyObject` form, which is guaranteed to be hashable through `NSObject`'s `-hash` and `-isEqual:`. I'll take a closer look!

> Also, don't forget NSSet?

`NSSet` already has a similar implementation of `_HasCustomAnyHashableRepresentation` elsewhere in the Foundation overlay:

```swift
extension NSSet : _HasCustomAnyHashableRepresentation {
  // Must be @nonobjc to avoid infinite recursion during bridging
  @nonobjc
  public func _toCustomAnyHashable() -> AnyHashable? {
    return AnyHashable(self as! Set<AnyHashable>)
  }
}
```

(`Set` has been `Hashable` from the start -- it did not need to wait for conditional conformances with that, because its contents are always guaranteed to be hashable.)

Reply:
So apparently the same issue applies to `NSSet` and `NSDictionary` keys -- and it means that there are `NSSet`/`NSDictionary` instances that can't be bridged to Swift. 🙊

https://bugs.swift.org/browse/SR-7284

Reply:
@jrose-apple It seems this opened a big can of worms. `_SwiftValue` instances for non-`Hashable` values spectacularly break language assumptions on how bridging interacts with `AnyHashable` not just in these new cases, but for existing ones as well.

I suggest to land this PR as is, and to track the `_SwiftValue` vs `AnyHashable` issue separately, because it reaches far outside the scope of this PR. 

I suspect these `_toCustomAnyHashable` implementations (including the pre-existing one for `NSSet`) are actually fine -- the problem is that `self as? Set<AnyHashable>` not only fails sometimes, but unexpectedly, it also crashes.

Reply:
I'm less interested in non-hashable *keys* than I am in non-hashable *values,* but maybe you're right that this isn't important in practice, since all NSObjects are Hashable.

Reply:
cc @michael-lehew in case he has any additional thoughts here.

Reply:
I'd be very curious to follow on to the tracked `_SwiftValue` vs `AnyHashable` issue.  I agree with @lorentey that it can be considered separately, but landing this PR really forces the issue, especially as more non-ObjC types (thinking model-layer) find their way into the value side of  dictionary.

Reply:
Okay! My plan is to make sure everything that can be put in an `NSSet` and `NSDictionary` can be converted to `AnyHashable`. This includes Swift types that aren't natively `Hashable`, since their NSObject boxes do implement hashing. Let's move on to [SR-7284](https://bugs.swift.org/browse/SR-7284) to fix the underlying issue.

Note that while the issue is serious, it is a subtle corner case that does not normally occur in applications -- and if it does, users have an easy source-level workaround. (By simply implementing `Hashable`.) The additional wrinkles introduced by this PR require even more convoluted Swift->Objective-C->Swift bridging scenarios, and I believe they are even less likely to occur in practice.